database/schema.sql

```diff
@@ -0,0 +1,167 @@
+
+-- Valid for squlite3 3.10+ (if not exists, without rowid).
+
+-- As sql keywords are composite and have spaces inside them,
+-- I used 2 spaces to separate names, types and constraints visually.
+--
+-- Also, sqlite is said to create indices on PKs automatically, so I will omit them.
+--
+-- As the sqlite will create additional `rowid` unique PK even if another
+--  non integer or integer desc PK is present, some of tables are created
+--  WITHOUT ROWID.
+
+-- Creating 'Courses' table.
+--
+create table if not exists Courses (
+
+    -- Sqlite3 will force ascending primary key to be non-null, replacing
+    --  null with autoincremented key (while sqlite2 won't).
+    --
+    id    INTEGER,
+    desc  TEXT     null,
+
+    primary key (id asc)
+);
+
+-- Creating 'Subjects' table.
+--
+create table if not exists Subjects (
+    id         INTEGER  not null,
+    course_id  INTEGER  not null,
+    desc       TEXT     null,
+
+    primary key (id, course_id),
+    foreign key (course_id) references Courses (id)
+);
+
+create index if not exists Subject_course_ids on Subjects (course_id);
+
+-- Creating 'Students' table.
+--
+create table if not exists Students (
+    -- The "primary key" doesn't make field non null in sqlite.
+    --
+    addr  BLOB  not null,
+
+    primary key (addr)
+
+) without rowid;
+
+-- Creating 'StudentCourses' table.
+--
+-- [Appears to be] many-to-many ref between Students and Courses.
+--
+create table if not exists StudentCourses (
+    student_addr  BLOB     not null,
+    course_id     INTEGER  not null,
+
+    -- Composite primary.
+    --
+    primary key (student_addr, course_id),
+
+    foreign key (student_addr) references Students (addr),
+    foreign key (course_id)    references Courses  (id)
+
+) without rowid;
+
+create index if not exists StudentCourses_student_addr on StudentCourses (student_addr);
```

Comment:
I don't know squlite that well. What kind of index do you get here as default? Is the default index strategy good for BLOB data?

Reply:
@lambda76 SQLite only uses btree indices, so we have no choice here. It would be better to have hashtable indices, but we'd need something like PostgreSQL for that. We don't expect high load on the Educator database, so we resolve this tradeoff in favor of portability and simplicity by using SQLite.